### PR TITLE
Add differentiable forward kinematics

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/embodied/robot/forward_kinematics.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            jac = jposes.new_zeros(angles.shape[0], 6, robot.dof)
+            sel = robot.links[link_id].angle_ids
+            jac[:, :, sel] = se3.adjoint(se3.inverse(pose)) @ jposes[:, :, sel]
+            jacs.append(jac)
+
+        return jacs, rets
+
+    class ForwardKinematics(torch.autograd.Function):
+        @classmethod
+        def forward(cls, ctx, angles):
```
Comment (cpaxton) [inline theseus/embodied/robot/forward_kinematics.py:97]: What is ctx? maybe a comment?

File: theseus/embodied/robot/link.py [hunk 2]
```diff
@@ -60,6 +65,9 @@ def set_children(self, children: List[Any]):
     def set_ancestors(self, ancesotrs: List["Link"]):
         self._ancestors = ancesotrs
 
+    def set_angle_ids(self, angle_ids: List[int]):
```
Comment (cpaxton) [inline theseus/embodied/robot/link.py:68]: what is this supposed to do?

File: theseus/embodied/robot/link.py [hunk 3]
```diff
@@ -44,6 +45,10 @@ def children(self) -> List[Any]:
     def ancestors(self) -> List["Link"]:
         return self._ancestors
 
+    @property
+    def angle_ids(self):
```
Comment (cpaxton) [inline theseus/embodied/robot/link.py:49]: What is this supposed to do?

File: theseus/embodied/robot/forward_kinematics.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                    )
+                ).unsqueeze(-1)
+                grad_input[:, angle_ids] += (
+                    ctx.jposes[:, :, angle_ids].transpose(1, 2) @ temp
+                ).squeeze(-1)
+
+            return grad_input
+
+    return (
+        ForwardKinematics,
```
Comment (cpaxton) [inline theseus/embodied/robot/forward_kinematics.py:126]: this does feel a bit clunky... although i guess we have the wrapper fn below

File: theseus/embodied/robot/forward_kinematics.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        _forward_kinematics_impl,
+        _jforward_kinematics_impl,
+        _forward_kinematics_helper,
+        _jforward_kinematics_helper,
+    )
+
+
+def get_forward_kinematics(robot: Robot, link_names: Optional[List[str]] = None):
```
Comment (cpaxton) [inline theseus/embodied/robot/forward_kinematics.py:134]: why just these 2?

**Review by cpaxton — COMMENTED** [review]
lgtm in general, although i am a bit confused about some things

File: theseus/embodied/robot/link.py (hunk 2 above)
Comment (fantaosha) [reply theseus/embodied/robot/link.py:68]: This is the ids of non-fixed joints that a link depends on. For example, a 7-link robot arm, the second link only depends on joint 1 and joint2. This is used to improve the computational efficiency of jacobians.

File: theseus/embodied/robot/forward_kinematics.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+from theseus.geometry.functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicsFactory(robot: Robot, link_names: Optional[List[str]] = None):
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:19]: The outputs are forward kinematics for the given links of a specific robot. In some problems, we don't have to run the full forward kinematics if only interested in a subset of links.

File: theseus/embodied/robot/forward_kinematics.py (hunk 5 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:134]: That's correct. The other outputs are only used for tests.

File: theseus/embodied/robot/forward_kinematics.py (hunk 1 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:97]: `ctx` is used in PyTorch for analytic backward propagation. Here we present analytical backward function for forward kinematics, which yields 10x speedup.

File: theseus/embodied/robot/forward_kinematics.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
```
Comment (luisenp) [inline theseus/embodied/robot/forward_kinematics.py:23]: Maybe do `links = robot.links if link_names is None else [robot.link_map[name] for name in link_names]`

Also, I think this is better as a method of `Robot` class. Something like 
```python
def get_links(self, names: Optional[List[str]]) -> List[Link]:
```

File: theseus/embodied/robot/forward_kinematics.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    for link in links:
+        ancestors += [anc for anc in link.ancestors]
+        joint_ids += link.angle_ids
+    pose_ids = sorted(list(set([anc.id for anc in ancestors] + link_ids)))
+    joint_ids = sorted(list(set(joint_ids)))
+
+    def _forward_kinematics_helper(angles: torch.Tensor):
+        if angles.ndim != 2 or angles.shape[1] != robot.dof:
+            raise ValueError(
+                f"Joint angles for {robot.name} should be {robot.dof}-D vectors"
```
Comment (luisenp) [inline theseus/embodied/robot/forward_kinematics.py:40]: Maybe add "{robot.dof}-D vectors with a batch dimension."?

File: theseus/embodied/robot/forward_kinematics.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            )
+
+        poses: List[Optional[torch.Tensor]] = [None] * robot.num_links
+        poses[0] = angles.new_zeros(angles.shape[0], 3, 4)
```
Comment (luisenp) [inline theseus/embodied/robot/forward_kinematics.py:44]: Replace with `se3.new_identity(angles)`.

File: theseus/embodied/robot/forward_kinematics.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    def _jforward_kinematics_helper(
+        poses: List[Optional[torch.Tensor]],
+    ) -> torch.Tensor:
+        jposes = poses[0].new_zeros(poses[0].shape[0], 6, robot.dof)
+
+        for id in pose_ids[1:]:
+            link: Link = robot.links[id]
+            joint: Joint = link.parent
+            if joint.id >= robot.dof:
```
Comment (luisenp) [inline theseus/embodied/robot/forward_kinematics.py:74]: Is this check possible because joints are assumed to be 1D?

File: theseus/embodied/robot/forward_kinematics.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        ForwardKinematics,
+        _forward_kinematics_impl,
+        _jforward_kinematics_impl,
+        _forward_kinematics_helper,
+        _jforward_kinematics_helper,
+    )
+
+
+def get_forward_kinematics(robot: Robot, link_names: Optional[List[str]] = None):
+    ForwardKinematics, _, jforward_kinematics, _, _ = ForwardKinematicsFactory(
```
Comment (luisenp) [inline theseus/embodied/robot/forward_kinematics.py:142]: Rename output as `ForwardKinematics` --> `fk_autograd_fn_cls`?

File: theseus/embodied/robot/robot.py [hunk 12]
```diff
@@ -51,11 +51,18 @@ def get_origin(urdf_origin: Optional[urdf.Pose] = None) -> torch.Tensor:
                 origin[:, :, 3] = torch.tensor(urdf_origin.xyz, dtype=dtype)
 
             if urdf_origin.rpy is not None:
-                rpy = urdf_origin.rpy
-                rot_x = so3.exp(torch.tensor([[rpy[0], 0, 0]], dtype=dtype))
-                rot_y = so3.exp(torch.tensor([[0, rpy[1], 0]], dtype=dtype))
-                rot_z = so3.exp(torch.tensor([[0, 0, rpy[2]]], dtype=dtype))
-                origin[:, :, :3] = rot_x @ rot_y @ rot_z
+                rpy = torch.tensor(urdf_origin.rpy, dtype=dtype)
```
Comment (luisenp) [inline theseus/embodied/robot/robot.py:54]: Was this rewritten for efficiency? (maybe to avoid matrix products?)

**Review by luisenp — CHANGES_REQUESTED** [review]
Great job adding all this stuff! I left some comments to address.

File: theseus/embodied/robot/link.py (hunk 2 above)
Comment (exhaustin) [reply theseus/embodied/robot/link.py:68]: Compared to tracing manually back to the root, the computational complexity isn't really affected by caching the dependent joints, since you still need linear time to go through them.
Also, performance shouldn't be an issue since you have a factory pattern to generate an object that has all of the computational paths traced.

File: theseus/embodied/robot/forward_kinematics.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+    link_ids: List[int] = [link.id for link in links]
+
+    ancestors: List[Link] = []
+    joint_ids: List[int] = []
+    for link in links:
+        ancestors += [anc for anc in link.ancestors]
+        joint_ids += link.angle_ids
+    pose_ids = sorted(list(set([anc.id for anc in ancestors] + link_ids)))
+    joint_ids = sorted(list(set(joint_ids)))
```
Comment (exhaustin) [inline theseus/embodied/robot/forward_kinematics.py:35]: Same comment as https://github.com/facebookresearch/theseus/pull/417#discussion_r1091528884

File: theseus/embodied/robot/joint.py [hunk 14]
```diff
@@ -114,7 +114,7 @@ def dof(self) -> int:
         return 0
 
     def relative_pose(self, *args) -> torch.Tensor:
-        if len(args) != 1:
+        if len(args) != 0:
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/joint.py:117]: Yeah, I just found this bug in #420 and it is a little complicated to fix it in #417 :)

File: theseus/embodied/robot/forward_kinematics.py (hunk 13 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:35]: - Yeah, I will refactor the code with `link.joint` and `link.parent`.
- I think keeping the ancestor ids is important in lots of applications. This makes it possible to get the sparsity structure when the robot model is used in collision.
- We have `robot.joints` and the users get the joint names by `robot.joints[joint_id].name`.

File: theseus/embodied/robot/forward_kinematics.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        jposes = poses[0].new_zeros(poses[0].shape[0], 6, robot.dof)
+
+        for id in pose_ids[1:]:
+            link: Link = robot.links[id]
+            joint: Joint = link.parent
+            if joint.id >= robot.dof:
+                break
+            jposes[:, :, joint.id : joint.id + 1] = (
+                se3.adjoint(poses[link.id]) @ joint.axis
+            )
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:78]: This code applies all types of joints, including both revolute and prismatic.

File: theseus/embodied/robot/robot.py [hunk 16]
```diff
@@ -135,6 +142,12 @@ def get_origin(urdf_origin: Optional[urdf.Pose] = None) -> torch.Tensor:
         for link in robot.links:
             if link.parent is not None:
                 link.set_ancestors(link.parent.parent.ancestors + [link.parent.parent])
+                angle_ids = (
+                    link.parent.parent.angle_ids
+                    if isinstance(link.parent, FixedJoint)
+                    else link.parent.parent.angle_ids + [link.parent.id]
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/robot.py]: Added `Link.parent_link`, and renamed some properties of a link, i.e., `Link.parent->Link.parent_joint`, `Link.children->Link.child_joints`, `Link.angle_ids -> Link.ancestor_active_joint_ids`, etc.

File: tests/embodied/robot/test_forward_kinematics.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+    rng = torch.Generator()
+    rng.manual_seed(0)
+    angles = torch.rand(batch_size, robot.dof, generator=rng, dtype=dtype)
+
+    jacs_impl = torch.autograd.functional.jacobian(fkin_impl, angles, vectorize=True)
+    jacs = torch.autograd.functional.jacobian(fkin, angles, vectorize=True)
+
+    for jac, jac_impl in zip(jacs, jacs_impl):
+        assert torch.allclose(jac_impl, jac, atol=TEST_EPS)
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply tests/embodied/robot/test_forward_kinematics.py:38]: Replaced.

File: theseus/embodied/robot/robot.py (hunk 12 above)
Comment (fantaosha) [reply theseus/embodied/robot/robot.py]: That's correct.

File: theseus/embodied/robot/forward_kinematics.py (hunk 10 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:74]: Yeah, that's our current assumption.

File: theseus/embodied/robot/forward_kinematics.py (hunk 8 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:40]: Not sure. If I remember correctly, we only consider the second dimension in our Lie group code.

File: theseus/embodied/robot/link.py (hunk 2 above)
Comment (fantaosha) [reply theseus/embodied/robot/link.py]: > We should have a more explicit/descriptive name for this than `angle_ids`.

Now `angle_ids -> ancestor_active_joint_ids`.

File: theseus/embodied/robot/forward_kinematics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:19]: Is `->Any` OK?

File: tests/embodied/robot/test_forward_kinematics.py [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+
+URDF_REL_PATH = "../kinematics/data/panda_no_gripper.urdf"
+urdf_path = os.path.join(os.path.dirname(__file__), URDF_REL_PATH)
+
+
+@pytest.mark.parametrize("batch_size", [1, 20, 40])
+@pytest.mark.parametrize("dtype", [torch.float32, torch.float64])
+def test_exp(batch_size: int, dtype: torch.dtype):
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply tests/embodied/robot/test_forward_kinematics.py]: Fixed now.

File: theseus/embodied/robot/forward_kinematics.py [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+from typing import List, Optional
+
+from theseus.labs.lie_functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+
+# TODO: Add support for joints with DOF>1
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:16]: #473

File: theseus/embodied/robot/forward_kinematics.py (hunk 7 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:23]: refactored

File: theseus/embodied/robot/forward_kinematics.py (hunk 9 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:38]: Maybe we need more discussions about this the behavior of `se3.new_identity()`. For example, how to specify the dimensions.

File: theseus/embodied/robot/forward_kinematics.py (hunk 8 above)
Comment (luisenp) [reply theseus/embodied/robot/forward_kinematics.py:34]: I'm just suggesting a more descriptive error message for what your code is already doing (you are checking ndim == 2 and dof being in dim 1). Your current message implies that `ndim` should be `==1`.

File: theseus/embodied/robot/forward_kinematics.py (hunk 9 above)
Comment (luisenp) [reply theseus/embodied/robot/forward_kinematics.py:38]: Is there any reasonable option other than the code below? 

```python
g = se3.new_identity(batch_size)
# g.shape = (batch_size, 3, 4)
```

File: theseus/embodied/robot/forward_kinematics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/embodied/robot/forward_kinematics.py:19]: Why `Any`? What possible options can this return?

File: theseus/embodied/robot/forward_kinematics.py (hunk 11 above)
Comment (luisenp) [reply theseus/embodied/robot/forward_kinematics.py:138]: Is there anything problematic about this suggestion? `CamelCase` is usually used for class definitions.

File: theseus/embodied/robot/forward_kinematics.py (hunk 9 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:38]: The `liegroup` has no `new_identity()` method. This is definitely in our TODO list and I will go back to this once `new_identity()` is added to Lie group.

File: theseus/embodied/robot/forward_kinematics.py (hunk 11 above)
Comment (fantaosha) [reply theseus/embodied/robot/forward_kinematics.py:138]: I'm OK for either. I will address this comment after all the PRs have been merged since lots of functions use `ForwardKinematics`.
